examples/mixture_models/dirichlet_mixture_of_multinomials.ipynb

```diff
@@ -0,0 +1,1815 @@
+{
```

Comment:
<ul><li>By n, you mean n_samples here, right? If yes, would be better to use this denomination, to match the one in the code</li><li>"While either of these would probably overcome the problems with the default sampler": why is so?</li><li>"here we'll instead switch to a standard, Metropolis step method, increasing our number of draws to overcome the inefficiency relative to NUTS": this formulation is problematic to me, as the problem doesn't come from NUTS, it comes from the <em>model</em>. Switching to Metropolis doesn't make the model more efficient; it just makes the warnings go away, both because Metropolis just doesn't have them, and because the algorithm <em>itselft</em> is less robust. The current framing implies that switching to Metropolis makes the convergence problems <em>go away</em> when it just <em>hides</em> them. As it's a mistake that's often made by people coming to Bayes, I think we should change the formulation above. What do you think?</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/18/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Reply:
<p>I haven't drilled down into the NUTS sampling errors, but I'm guessing that we're getting logp's or dlogp's of&nbsp;<code>-inf</code>&nbsp;due to the large deviation from the multinomial predictive density (e.g.&nbsp;<code>observed_counts[3, 1] == 1</code>&nbsp;is pretty improbable given that the expectation is 15).</p><p> </p><blockquote>Switching to Metropolis doesn't make the model more efficient; it just makes the warnings go away, both because Metropolis just doesn't have them, and because the algorithm&nbsp;<em>itselft</em>&nbsp;is less robust. The current framing implies that switching to Metropolis makes the convergence problems&nbsp;<em>go away</em>&nbsp;when it just&nbsp;<em>hides</em>&nbsp;them.</blockquote><p> </p><p>Since Metropolis either (A) rejects jumps into regions of the parameter space where logp is&nbsp;<code>-inf</code>, or (B) doesn't even know about the numerically unstable gradient, it doesn't fail the way NUTS does.</p><p> </p><blockquote>"While either of these would probably overcome the problems with the default sampler"</blockquote><p> </p><p>Haven't actually tried it, but I'm conjecturing here that the <s>Bernoulli</s> categorical likelihood elements (note: Bernoulli was a mistake) won't be as "surprised" (so to say) with hits to the low frequency species since any one observation is fairly believable. Equivalently with the collapsed likelihood, since there's nothing surprising about any combination of species counts. (...Does this explanation make any sense?)</p><p> </p><p>(Addendum: I went ahead and tried out the collapsed model, which works the way I expected.)</p><p> </p><blockquote>this formulation is problematic to me [...] As it's a mistake that's often made by people coming to Bayes, I think we should change the formulation above.</blockquote><p> </p><p>Do you have a suggestion for better wording?</p><p> </p><blockquote>What do you think?</blockquote><p> </p><p>Mmm...I'm pretty okay with mostly sweeping these issues under the rug in this example. I think the details you're considering here should be in an e.g. "Diagnosing Biased Inference" or "Reparameterizing Problematic Models" notebooks. Here that problem serves as good motivation for using a better model, but I don't think we need to deal with it much more than that.</p>

Reply:
<p>I agree that it might be a bit out of the scope of this Notebook to explore why NUTS fails. Maybe just warn more explicitly that this is not a real solution to the problem, but just a shortcut so you have something to compare with the proper DM model?</p>

Reply:
<blockquote>I'm guessing that we're getting logp's or dlogp's of&nbsp;<code>-inf</code>&nbsp;due to the large deviation from the multinomial predictive density (e.g.&nbsp;<code>observed_counts[3, 1] == 1</code>&nbsp;is pretty improbable given that the expectation is 15).</blockquote><p>Yeah, that'd be my guess too. We could probably help NUTS run by selecting better priors than the current uniform ones, but, by design, the issue is deeper -- the model is just not a good enough approximation of the generative process.</p><p> </p><blockquote>Does this explanation make any sense?</blockquote><p>Yeah thanks, I think I understand what you mean. You can see if you can briefly add that in the text above, but it's not a blocker.</p><p> </p><blockquote>'m pretty okay with mostly sweeping these issues under the rug in this example. [...] Here that problem serves as good motivation for using a better model, but I don't think we need to deal with it much more than that.</blockquote><p>I agree with that. What I'm saying is we should say that doing so is not <em>fixing</em> issues but <em>sweeping</em> them under the rug, otherwise readers could think that just switching to Metropolis when NUTS breaks is a valid workflow, whereas the latter happening should be a red alert -- actually, I think there'd be value in mentioning or even showing that NUTS doesn't even want to sample such a model --&gt; folk theorem must pop in our head.</p><p> </p><blockquote>Do you have a suggestion for better wording?</blockquote><p>The way you reworded it is already quite good! I'd just add a bit to it. Something like:</p><p>"... here we'll instead switch to the Metropolis step method, which ignores some of the geometric pathologies of our naïve model. This is important: switching to Metropolis does not not <em>fix </em>our model's issues, rather it <em>sweeps them under the rug</em>. In fact, if you try running this model with PyMC's default sampler, NUTS will break loudly from the beginning. When that happens, this should be a red alert that there is something wrong in our model.</p><p>You'll also notice below that we have to increase considerably the number of draws we take from the posterior, because Metropolis is much less efficient than NUTS."</p><p> </p><p>Ricardo's last comment is also spot-on:</p><blockquote>And later don't emphasize so much that the diagnostics look nice, but caution again that while they may look nicer we are probably glossing over very real problems in the posterior approximation that the NUTS made clear.</blockquote><p> </p><p> </p>